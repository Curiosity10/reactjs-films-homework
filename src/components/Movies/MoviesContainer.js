import React, { useLayoutEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Movies from './Movies';
import {
  fetchMoviesByFilter,
  fetchSearchMovies,
  fetchMoviesByGenre,
} from '../../modules/app/appActions';
import layoutSelector from '../../modules/ui/uiSelectors';
import {
  filterSelector,
  currentPageSelector,
  isLoadingSelector,
  genreIdSelector,
  searchQuerySelector,
  hasMorePagesSelector,
  videoSrcSelector,
  moviesSelector,
  errorMessageSelector,
} from '../../modules/app/appSelectors';
import { MAX_SCROLL_HEIGHT } from './utils/constants';

export const handleScroll = ({
  isLoading,
  hasMorePages,
  filter,
  getMoviesByFilter,
  getMoviesByGenre,
  getSearchMovies,
  genreId,
  searchQuery,
  window,
}) => {
  if (isLoading) {
    return;
  }

  const { scrollHeight, clientHeight } = window.document.documentElement;
  const scrollPos = window.pageYOffset;
  const currentPosition = scrollPos + clientHeight;
  const percentageScrolled = currentPosition / scrollHeight;

  if (percentageScrolled > MAX_SCROLL_HEIGHT && hasMorePages) {
    switch (filter) {
      case 'popular':
      case 'top_rated':
      case 'upcoming':
        getMoviesByFilter();
        break;
      case 'Genres':
        getMoviesByGenre(genreId);
        break;
      case 'Search':
        getSearchMovies(searchQuery);
        break;
      default:
        break;
    }
  }
};
const MovieContainer = ({
  isLoading, filter, currentPage, genreId,
  searchQuery, movies, layout, getMoviesByFilter,
  getSearchMovies, getMoviesByGenre, videoSrc,
  hasMorePages, errorMsg,
}) => {
  useLayoutEffect(() => {
    const callback = () => handleScroll({
      isLoading,
      hasMorePages,
      filter,
      getMoviesByFilter,
      getMoviesByGenre,
      getSearchMovies,
      genreId,
      searchQuery,
      window,
    });

    document.addEventListener('scroll', callback);

    return () => {
      document.removeEventListener('scroll', callback);
    };
  }, [
    isLoading, hasMorePages, filter,
    getMoviesByFilter, getMoviesByGenre, genreId,
    getSearchMovies, searchQuery,
  ]);

  return (
    <Movies
      errorMsg={errorMsg}
      page={currentPage}
      videoSrc={videoSrc}
      isLoading={isLoading}
      movies={movies}
      layout={layout}
    />
  );
};

const mapStateToProps = state => ({
  movies: moviesSelector(state),
  layout: layoutSelector(state),
  currentPage: currentPageSelector(state),
  isLoading: isLoadingSelector(state),
  filter: filterSelector(state),
  genreId: genreIdSelector(state),
  searchQuery: searchQuerySelector(state),
  hasMorePages: hasMorePagesSelector(state),
  videoSrc: videoSrcSelector(state),
  errorMsg: errorMessageSelector(state),
});

const actions = {
  getMoviesByFilter: fetchMoviesByFilter,
  getSearchMovies: fetchSearchMovies,
  getMoviesByGenre: fetchMoviesByGenre,
};

MovieContainer.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  layout: PropTypes.string.isRequired,
  currentPage: PropTypes.number.isRequired,
  filter: PropTypes.string.isRequired,
  genreId: PropTypes.string,
  searchQuery: PropTypes.string,
  videoSrc: PropTypes.string,
  errorMsg: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  hasMorePages: PropTypes.bool.isRequired,
  movies: PropTypes.arrayOf(PropTypes.object),
  getMoviesByFilter: PropTypes.func.isRequired,
  getSearchMovies: PropTypes.func.isRequired,
  getMoviesByGenre: PropTypes.func.isRequired,
};

MovieContainer.defaultProps = {
  genreId: null,
  searchQuery: '',
  videoSrc: '',
  movies: [],
  errorMsg: '',
};

export default connect(mapStateToProps, actions)(MovieContainer);
