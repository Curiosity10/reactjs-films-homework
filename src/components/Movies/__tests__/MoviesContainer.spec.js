import React from 'react';
import { create, act } from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import ReactDOM from 'react-dom';
import { JSDOM } from 'jsdom';
import MoviesContainer from '..';
import { handleScroll } from '../MoviesContainer';
import { genres } from '../../../__mocks__/genres.json';

global.fetch = require('jest-fetch-mock');

describe('MoviesContainer component renders correctly', () => {
  beforeAll(() => {
    const dom = new JSDOM();
    global.document = dom.window.document;
    global.window = dom.window;
    ReactDOM.createPortal = jest.fn(element => element);
  });

  afterEach(() => {
    ReactDOM.createPortal.mockClear();
  });

  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  const movies = [
    {
      title: 'test',
      id: 1,
      genre_ids: [
        {
          id: 12,
          name: 'Action',
        },
        {
          id: 16,
          name: 'Adventure',
        },
      ],
      overview: 'desc',
    },
    {
      title: 'test2',
      id: 2,
      genre_ids: [
        {
          id: 21,
          name: 'Action',
        },
        {
          id: 23,
          name: 'Adventure',
        },
      ],
      overview: 'desc2',
    },
  ];

  it('MoviesContainer grid render, unmount and scroll correctly', () => {
    const initialState = {
      app: {
        movies,
        genres,
        isLoading: true,
        currentPage: 1,
        error: false,
        filter: 'Trending',
        genreId: null,
        searchQuery: '',
        videoKey: '',
        pagesCount: 1,
      },
      ui: {
        layout: 'grid',
      },
    };
    const store = mockStore(initialState);
    let tree;

    act(() => {
      tree = create(
        <Provider store={store}>
          <MoviesContainer />
        </Provider>,
      );
    });

    const event = new Event('scroll', {});
    act(() => {
      document.dispatchEvent(event);
    });

    expect(tree.toJSON()).toMatchSnapshot();
    tree.unmount();
  });

  it('MoviesContainer table render correctly', () => {
    const initialState = {
      app: {
        movies,
        genres,
        isLoading: true,
        currentPage: 1,
        error: false,
        filter: 'Trending',
        genreId: null,
        searchQuery: '',
        videoKey: '',
        pagesCount: 1,
      },
      ui: {
        layout: 'table',
      },
    };
    const store = mockStore(initialState);

    const tree = create(
      <Provider store={store}>
        <MoviesContainer />
      </Provider>,
    );

    const button = tree.root.findAllByProps({ 'aria-label': 'Watch Now' })[0];
    const mockFn = jest.fn(button.props.onClick);

    act(() => {
      mockFn();
    });

    expect(mockFn).toHaveBeenCalled();
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('MoviesContainer without layout renders correctly', () => {
    const initialState = {
      app: {
        movies,
        genres,
        isLoading: true,
        currentPage: 1,
        error: false,
        filter: 'Trending',
        genreId: null,
        searchQuery: '',
        videoKey: '',
        pagesCount: 1,
      },
      ui: {
        layout: '',
      },
    };
    const store = mockStore(initialState);

    const tree = create(
      <Provider store={store}>
        <MoviesContainer />
      </Provider>,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('MoviesContainer without movies table render correctly', () => {
    const initialState = {
      app: {
        movies: [],
        genres,
        isLoading: true,
        currentPage: 1,
        error: false,
        filter: 'Trending',
        genreId: null,
        searchQuery: '',
        videoKey: '',
        pagesCount: 1,
      },
      ui: {
        layout: 'table',
      },
    };
    const store = mockStore(initialState);

    let tree;
    act(() => {
      tree = create(
        <Provider store={store}>
          <MoviesContainer />
        </Provider>,
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});

describe('Fetch on scroll work correctly', () => {
  const currentPage = 1;
  const genreId = 12;
  const searchQuery = '123';

  const window = {
    document: {
      documentElement: {
        scrollHeight: 1600,
        clientHeight: 1200,
      },
    },
    pageYOffset: 500,
  };

  const getMoviesByFilter = jest.fn();
  const getMoviesByGenre = jest.fn();
  const getSearchMovies = jest.fn();

  it('Fetch on scroll if not load and filter trending', () => {
    const isLoading = false;
    const hasMorePages = true;
    const filter = 'popular';
    handleScroll({
      isLoading,
      hasMorePages,
      filter,
      currentPage,
      getMoviesByFilter,
      getMoviesByGenre,
      genreId,
      getSearchMovies,
      searchQuery,
      window,
    });
    expect(getMoviesByFilter).toHaveBeenCalled();
  });

  it('Fetch on scroll if not load and filter top rated', () => {
    const hasMorePages = true;
    const isLoading = false;
    const filter = 'top_rated';
    handleScroll({
      isLoading,
      hasMorePages,
      filter,
      currentPage,
      getMoviesByFilter,
      getMoviesByGenre,
      genreId,
      getSearchMovies,
      searchQuery,
      window,
    });
    expect(getMoviesByFilter).toHaveBeenCalled();
  });

  it('Fetch on scroll if not load and filter coming soon', () => {
    const hasMorePages = true;
    const isLoading = false;
    const filter = 'upcoming';
    handleScroll({
      isLoading,
      hasMorePages,
      filter,
      currentPage,
      getMoviesByFilter,
      getMoviesByGenre,
      genreId,
      getSearchMovies,
      searchQuery,
      window,
    });
    expect(getMoviesByFilter).toHaveBeenCalled();
  });

  it('Fetch on scroll if not load and filter genres', () => {
    const hasMorePages = true;
    const isLoading = false;
    const filter = 'Genres';
    handleScroll({
      isLoading,
      hasMorePages,
      filter,
      currentPage,
      getMoviesByFilter,
      getMoviesByGenre,
      genreId,
      getSearchMovies,
      searchQuery,
      window,
    });
    expect(getMoviesByGenre).toHaveBeenCalled();
  });

  it('Fetch on scroll if not load and filter search', () => {
    const hasMorePages = true;
    const isLoading = false;
    const filter = 'Search';
    handleScroll({
      isLoading,
      hasMorePages,
      filter,
      currentPage,
      getMoviesByFilter,
      getMoviesByGenre,
      genreId,
      getSearchMovies,
      searchQuery,
      window,
    });
    expect(getSearchMovies).toHaveBeenCalled();
  });

  it('Fetch on scroll if not load without filter', () => {
    const hasMorePages = true;
    const isLoading = false;
    const filter = 'Hello';
    const mockFn = jest.fn(() => handleScroll({
      isLoading,
      hasMorePages,
      filter,
      currentPage,
      getMoviesByFilter,
      getMoviesByGenre,
      genreId,
      getSearchMovies,
      searchQuery,
      window,
    }));

    act(() => {
      mockFn();
    });

    expect(mockFn).toHaveBeenCalled();
  });

  it('Fetch on scroll if loading = true', () => {
    const hasMorePages = true;
    const isLoading = true;
    const filter = 'Hello';
    const mockFn = jest.fn(() => handleScroll({
      isLoading,
      hasMorePages,
      filter,
      currentPage,
      getMoviesByFilter,
      getMoviesByGenre,
      genreId,
      getSearchMovies,
      searchQuery,
      window,
    }));

    act(() => {
      mockFn();
    });

    expect(mockFn).toHaveBeenCalled();
  });

  it('Fetch on scroll if hasn\'t more pages', () => {
    const hasMorePages = false;
    const isLoading = false;
    const filter = 'Hello';
    const mockFn = jest.fn(() => handleScroll({
      isLoading,
      hasMorePages,
      filter,
      currentPage,
      getMoviesByFilter,
      getMoviesByGenre,
      genreId,
      getSearchMovies,
      searchQuery,
      window,
    }));

    act(() => {
      mockFn();
    });

    expect(mockFn).toHaveBeenCalled();
  });
});
