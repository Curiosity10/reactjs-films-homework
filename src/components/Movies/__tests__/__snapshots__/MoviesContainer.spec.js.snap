// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesContainer component renders correctly MoviesContainer grid render, unmount and scroll correctly 1`] = `
Array [
  <main
    className="container gridMovieList"
  >
    <article
      className="card"
    >
      <img
        alt="Poster"
        height="70%"
        src={
          Object {
            "process": [Function],
          }
        }
      />
      <div
        className="onHover"
      >
        <button
          aria-label="Watch"
          className="btnWatch"
          onClick={[Function]}
          type="button"
        >
          <span>
             Watch Now 
          </span>
        </button>
        <button
          aria-label="View Info"
          className="btn ghost"
          onClick={[Function]}
          type="button"
        >
          View Info
        </button>
      </div>
      <div
        className="info"
      >
        <div>
          <h3
            className="title"
            title="test"
          >
            test
          </h3>
          <p />
        </div>
      </div>
    </article>
    <article
      className="card"
    >
      <img
        alt="Poster"
        height="70%"
        src={
          Object {
            "process": [Function],
          }
        }
      />
      <div
        className="onHover"
      >
        <button
          aria-label="Watch"
          className="btnWatch"
          onClick={[Function]}
          type="button"
        >
          <span>
             Watch Now 
          </span>
        </button>
        <button
          aria-label="View Info"
          className="btn ghost"
          onClick={[Function]}
          type="button"
        >
          View Info
        </button>
      </div>
      <div
        className="info"
      >
        <div>
          <h3
            className="title"
            title="test2"
          >
            test2
          </h3>
          <p />
        </div>
      </div>
    </article>
  </main>,
  <div
    className="spinner"
  >
    <div
      className="ellipsis"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
    <p
      className="title"
    >
      Loading
    </p>
  </div>,
]
`;

exports[`MoviesContainer component renders correctly MoviesContainer table render correctly 1`] = `
Array [
  <main
    className="container tableMovieList"
  >
    <article
      className="card"
    >
      <img
        alt="Poster"
        src={
          Object {
            "process": [Function],
          }
        }
      />
      <div
        className="info"
      >
        <h2
          className="title"
        >
          test
        </h2>
        <p
          className="overview"
        >
          desc
        </p>
        <div
          className="rating"
        />
        <button
          aria-label="Watch Now"
          className="btn primary"
          onClick={[Function]}
          type="button"
        >
          Watch Now
        </button>
      </div>
    </article>
    <article
      className="card"
    >
      <img
        alt="Poster"
        src={
          Object {
            "process": [Function],
          }
        }
      />
      <div
        className="info"
      >
        <h2
          className="title"
        >
          test2
        </h2>
        <p
          className="overview"
        >
          desc2
        </p>
        <div
          className="rating"
        />
        <button
          aria-label="Watch Now"
          className="btn primary"
          onClick={[Function]}
          type="button"
        >
          Watch Now
        </button>
      </div>
    </article>
  </main>,
  <div
    className="videoContainer"
  >
    <div
      className="spinner"
    >
      <div
        className="ellipsis"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
      <p
        className="title"
      >
        Loading
      </p>
    </div>
  </div>,
  <div
    className="spinner"
  >
    <div
      className="ellipsis"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
    <p
      className="title"
    >
      Loading
    </p>
  </div>,
]
`;

exports[`MoviesContainer component renders correctly MoviesContainer without layout renders correctly 1`] = `
Array [
  <main
    className="container"
  />,
  <div
    className="spinner"
  >
    <div
      className="ellipsis"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
    <p
      className="title"
    >
      Loading
    </p>
  </div>,
]
`;

exports[`MoviesContainer component renders correctly MoviesContainer without movies table render correctly 1`] = `
Array [
  <main
    className="container"
  />,
  <div
    className="spinner"
  >
    <div
      className="ellipsis"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
    <p
      className="title"
    >
      Loading
    </p>
  </div>,
]
`;
